Test Components for Customizing and Saving Grid Views in Apps

labkey-danield opened the pull request:
#### Rationale
This is the test automation for customizing grids and saving views from app grids. It only impacts query grids, editable and and read only grids cannot be customized or save a view. There are several test component additions and changes.

This is a draft PR at this time. Test cases still need to be written, but I wanted to get a review of the components started before too many tests were written.

### Changes to QueryGrid and ResponsiveGrid.
For the grids there is now the Insert Column and Hide Column menu options. Clicking Insert Column will show a Customize Grid dialog. Clicking Hide Column immediately hides the column. The Edit Label functionality hasn't been implemented yet.

When a grid is updated a status message is shown in the panel header as well as the Undo and Save buttons. Clicking the Save button will show the Save View dialog.

<img width="1109" alt="Grid_Updates" src="https://user-images.githubusercontent.com/12738200/181841908-a96af7ad-d656-4f05-bd29-2edd3cb8f9b6.png">

If changes are made to a saved view the Save button will have a dropdown menu option that will allow the changed view to be saved with a different name:

<img width="595" alt="Screen Shot 2022-08-11 at 10 51 20 AM" src="https://user-images.githubusercontent.com/12738200/184203666-8d8798c0-274a-4bb2-a869-4f1535588a1c.png">

If a change is made to a saved view and the 'Save' button is clicked no Save View Dialog will be shown. The changes will be saved to the current view.

Changes to the default view will never show the "Save as..." option, the Save View Dialog will always be shown.

### Save View Dialog
Fairly straight forward. If the 'Make default view for all users' checkbox is checked this will become the default view for the grid, and the view name text box will be disabled.

<img width="662" alt="Grid_SaveView_Dialog" src="https://user-images.githubusercontent.com/12738200/181842080-44dcd7ed-0048-41d5-b401-ce6e2e6e851a.png">

### Manage Saved Views Dialog
Also fairly straight forward. This dialog will allow a user to rename or delete an exIsting view. The default view cannot be renamed or deleted. If you have changed the current (named) view the dialog will also not allow a user to edit or delete it:

<img width="667" alt="Screen Shot 2022-08-11 at 11 00 06 AM" src="https://user-images.githubusercontent.com/12738200/184208455-08cfb787-93b1-4981-8e32-d036e85613c9.png">

Clicking the pencil icon will allow the user to rename a view:

<img width="672" alt="Screen Shot 2022-08-11 at 11 00 54 AM" src="https://user-images.githubusercontent.com/12738200/184208622-2e1725fb-8d5c-4855-9b6e-9012cb4789fc.png">

Selecting a view for deletion gives a prompt and a Yes/No option:

<img width="672" alt="Screen Shot 2022-08-11 at 11 01 11 AM" src="https://user-images.githubusercontent.com/12738200/184208808-3d8c5975-9090-4914-aa0b-f5d211e949d8.png">

### Customize Grid Dialog
The Customize Grid dialog allows the addition of new fields/columns to the grid. This dialog appears when the Insert Column menu option is selected.

Fields are added from the Available Fields panel by clicking the "+" on the right hand side of the field name. Fields can be nested and are expanded\collapsed by clicking the "+" or "-" to the left of the field name.

The Shown in Grid panel shows field\column labels. The labels can be renamed or removed from this panel. Drag drop has not yet been implemented, and most likely will not be implemented.

<img width="995" alt="Grid_Customize_Dialog" src="https://user-images.githubusercontent.com/12738200/181842705-79b1e4a0-e576-4dd8-97ee-fc121d6cc016.png">

### View Menu on Grid
The view menu on the grid now has a "Customize Grid View", "Manage Saved Views" and a "Save Grid View" option. The "Customize" and "Save" options will show the appropriate dialog. The "Save Grid View" will always show the save view dialog regardless of the current view being a named view or not. It is equivalent to the "Save as..." button drop down menu.

<img width="313" alt="Screen Shot 2022-08-11 at 11 07 20 AM" src="https://user-images.githubusercontent.com/12738200/184209707-8a95aa5d-2fae-4521-ab8f-ac4de0f140b5.png">

#### Related Pull Requests
* https://github.com/LabKey/platform/pull/3598
* https://github.com/LabKey/sampleManagement/pull/1161
* https://github.com/LabKey/biologics/pull/1509

#### Changes
* Updated CustomizeView.saveCustomView to allow for unchecking the shared checkbox (LKS component).
* Added CustomizeGridModalDialog, SaveViewDialog and ManageViewDialog
* Refactored GirdPanelTest. Added a base test with shared methods.
* Added methods for header bar elements in QueryGrid (alert text, Save & Undo buttons and displayed view name).
* Added hideColumn and insertColumn methods to ResponsiveGrid.
* Added overload initColumnsAndIndices method so columnNames and indexes can be recalculated if needed.
* Removed view options from existing filter grid code. Views names are no longer shown like filter options.
* Added a GridPanelViewTest.

labkey-danield committed:
Merge branch 'develop' into fb_addViewTestsToGridPanelTest

labkey-danield committed:
Merge branch 'develop' into fb_addViewTestsToGridPanelTest

labkey-danield committed:
Merge branch 'develop' into fb_addViewTestsToGridPanelTest

labkey-danield committed:
Added CustomizeGridModal.
Updated CustomizeView.saveCustomView to allow for unchecking the shared checkbox.
Refactored GirdPanelTest. Added a base test with shared methods.
Added methods for header bar elements in QueryGrid (alert text, Save & Undo buttons).
Added SaveViewDialog to QueryGrid.
Added hideColumn and insertColumn methods to ResponsiveGrid.
Added overload initColumnsAndIndices method so columnNames and indexes can be recalculated if needed.

labkey-tchad commented on src/org/labkey/test/components/CustomizeView.java:
```diff
@@ -194,12 +194,20 @@ public void saveCustomView(String name, boolean shared, boolean inherit)
         {
             saveWindow.shareCheckbox.check();
         }
+        else
+        {
+            saveWindow.shareCheckbox.uncheck();
+        }
```
I'm worried that this will interfere with tests that depend on the current behavior (where `false` actually means to leave the default).
I'd suggest changing these flags to be nullable and update the other `saveCustomView` overloads to pass in `null` instead of false.
Then set the checkbox only if the flag isn't null.
Something like this:
```
public void saveCustomView(@Nullable String name) { saveCustomView(name, null, null); }

public void saveCustomView(String name, @Nullable Boolean shared) { saveCustomView(name, shared, null); }

public void saveCustomView(String name, @Nullable Boolean shared, @Nullable Boolean inherit)
{
    SaveWindow saveWindow = clickSave();

    if (shared != null)
    {
        saveWindow.shareCheckbox.set(shared);
    }

    if (inherit != null)
    {
        saveWindow.inheritCheckbox.set(inherit);
    }

etc.
```

labkey-danield replied:
Sounds reasonable. I updated the overloads and other tests that were currently passing in false for either of those fields.

labkey-tchad commented on src/org/labkey/test/components/ui/grids/CustomizeGridDialog.java:
```diff
@@ -0,0 +1,491 @@
+     */
+    public CustomizeGridDialog setShowAll(boolean checked)
+    {
+        if(checked)
+        {
+            elementCache().checkbox.check();
+        }
+        else
+        {
+            elementCache().checkbox.uncheck();
+        }
```
Simplify:
```suggestion
        elementCache().checkbox.set(checked);
```

labkey-danield replied:
Changed.

labkey-tchad commented on src/org/labkey/test/components/ui/grids/CustomizeGridDialog.java:
```diff
@@ -0,0 +1,491 @@
+     * @param fieldName The field to add.
+     * @return This dialog.
+     */
+    public CustomizeGridDialog addAvailableFieldToGrid(List<String> parentFields, String fieldName)
```
Combine both `addAvailableFieldToGrid` methods into one varargs method:
```suggestion
    public CustomizeGridDialog addAvailableFieldToGrid(String... fieldKeyParts)
```

labkey-danield replied:
Changed.

labkey-tchad commented on src/org/labkey/test/components/ui/grids/CustomizeGridDialog.java:
```diff
@@ -0,0 +1,491 @@
+     * @param parentFields The list of fields to expand.
+     * @return This dialog.
+     */
+    public CustomizeGridDialog expandAvailableFieldsTo(List<String> parentFields)
```
Combine both `expandAvailableField` methods into one varargs method:
```suggestion
    public CustomizeGridDialog expandAvailableFieldsTo(String... fieldKeyParts)
```

labkey-danield replied:
Done.

labkey-tchad commented on src/org/labkey/test/components/ui/grids/CustomizeGridDialog.java:
```diff
@@ -0,0 +1,491 @@
+        Assert.assertTrue(String.format("Field with data-fieldkey attribute '%s' is not visible in the 'Available Fields' panel.", fieldKey),
+                listItem.isDisplayed());
+
+        String iconClass;
+        if(expand)
+        {
+            iconClass = "fa-plus-square";
+        }
+        else
+        {
+            iconClass = "fa-minus-square";
+        }
```
Simplify:
```suggestion
        String iconClass = expand ? "fa-plus-square" : "fa-minus-square";
```

labkey-danield replied:
Done.

labkey-tchad commented on src/org/labkey/test/components/ui/grids/CustomizeGridDialog.java:
```diff
@@ -0,0 +1,491 @@
+        protected final WebElement fieldLabelEdit = Locator.tagWithClass("input", "form-control")
+                .refindWhenNeeded(shownInGridPanel());
+
+        protected WebElement availableFieldsPanel()
+        {
+            return contentPanelLocator.findElements(this).get(0);
+        }
+
+        protected WebElement shownInGridPanel()
+        {
+            return contentPanelLocator.findElements(this).get(1);
+        }
```
Are these sub-panels really that unstable that we can't cache them?
```suggestion
        protected final WebElement availableFieldsPanel = contentPanelLocator.index(0).findWhenNeeded(this);
        protected final WebElement shownInGridPanel = contentPanelLocator.index(1).findWhenNeeded(this);
```

labkey-danield replied:
No they should not be. I think this is a left over from a more generic approach I was thinking of taking but proved to be more complex that what was needed. Changed.

labkey-tchad commented on src/org/labkey/test/components/ui/grids/CustomizeGridDialog.java:
```diff
@@ -0,0 +1,491 @@
+
+        // The data-fieldkey attribute is only present in items in the Available Fields panel.
+        // Similar value to field-name (no spaces, but casing is the same). For child fields it will contain the parent path.
+        protected WebElement getListItemElementByFieldKey(WebElement panel, String fieldKey)
+        {
+            return Locator.tagWithClass("div", "list-group-item")
+                    .withAttribute("data-fieldkey", fieldKey)
+                    .findElement(panel);
+        }
```
Why specify the panel if this is only relevant to the available fields panel?
```suggestion
        protected WebElement getListItemElementByFieldKey(WebElement panel, String fieldKey)
        {
            return Locator.tagWithClass("div", "list-group-item")
                    .withAttribute("data-fieldkey", fieldKey)
                    .findElement(availableFieldsPanel());
        }
```

labkey-danield replied:
Agreed. Changed.

labkey-tchad commented on src/org/labkey/test/components/ui/grids/QueryGrid.java:
```diff
@@ -353,6 +402,13 @@ public List<FilterStatusValue> getFilterStatusFilterValues()
+        // The panel header element which will contain the Save and Undo buttons.
+        public WebElement panelHeader()
+        {
+            return getComponentElement().findElement(By.xpath("preceding-sibling::div[contains(@class,'panel-heading')]"));
```
Use our `Locator` class instead of the default `By`. We decorate `WebElements` with extra retry capabilities.
```suggestion
            return Locator.xpath("preceding-sibling::div[contains(@class,'panel-heading')]").findElement(this);
```

labkey-danield replied:
Done.

labkey-tchad commented on src/org/labkey/test/components/ui/grids/QueryGrid.java:
```diff
@@ -391,4 +447,104 @@ protected Locator locator()
+        /**
+         * Save the view.
+         */
+        public void saveView()
+        {
+            dismiss("Save", 1);
+        }
```
If you made this class non-static, you could wait for some state change in the `QueryGrid` when clicking save.
```suggestion
        public void saveView()
        {
            dismiss("Save", 1);
            // Wait for save button on the grid to disappear or something like that
            // e.g.
            waitFor(() -> !Locator.buttonContainingText("Save").findWhenNeeded(QueryGrid.this.getComponentElement()).isDisplayed(), 5_000);
        }
```

labkey-danield replied:
I could but it feels like that is something that would be more appropriate to be done in ModalDialog.dismiss(String buttonText).

labkey-danield replied:
Following up on this.
This code has changed. In the apps modifying a grid view and then saving it is different from LKS. In the app modifying the grid and saving the view are two different dialogs. Modifying the gird will result in an update to the grid, and that dialog does take a reference to the grid and waits for it to update. Saving a grid view does not result in a grid update and will not need to wait for it.

labkey-tchad commented on src/org/labkey/test/tests/component/GridPanelBaseTest.java:
```diff
@@ -0,0 +1,134 @@
+     * @param sampleType The sample type used to populate the grid.
+     * @return A queryGrid object.
+     */
+    protected QueryGrid initQueryGrid(String sampleType)
```
"init" usually implies configuring the server or test project. Maybe "goto" or "beginAt"?
```suggestion
    protected QueryGrid beginAtQueryGrid(String sampleType)
```

labkey-danield replied:
Fair enough. beginAt it is.

labkey-danield committed:
Merge branch 'develop' into fb_addViewTestsToGridPanelTest

labkey-danield committed:
Implemented code review feedback.
Added sanity test to make sure changes to test components are working as expected. Will remove it once the "real" tests are added.

labkey-danield committed:
Updated CustomizeView.saveCustomView to allow shared and inherited options to be unchecked. Passing in null will make no change to value.
Updated overloads of CustomizeView.saveCustomView to pass in null and not false.
Updated various tests to pass in null in place of false when calling CustomizeView.saveCustomView.

labkey-danield committed:
Add test annotation.

labkey-danield committed:
Added default view tests with filtering.
Modified locator for panel to account for text other than panel name in the header.
Fixed locator that searches for filter icons in grid column header.

cnathe committed:
Merge branch 'develop' into fb_addViewTestsToGridPanelTest

labkey-danield committed:
Merge branch 'develop' into fb_addViewTestsToGridPanelTest

labkey-danield committed:
Remove includeViews option from QueryGrid.getFilterStatusValues. View titles are in the panel title and no longer included with the filter pills.
Updated FilterStatusValue to only have filters.
Added methods to return text of filter pills from grid.
Updated expectation of clicking 'Save' button if in a view.
Added more tests to validate saving and editing grid view behavior.

labkey-danield committed:
Merge branch 'develop' into fb_addViewTestsToGridPanelTest

labkey-danield committed:
Merge branch 'develop' into fb_addViewTestsToGridPanelTest

labkey-danield committed:
Break out some dialog specific test cases.
Add a check to see if an available field has already been added.
Fix a bug when trying to expand a field that has already been expanded.

labkey-danield committed:
Merge branch 'develop' into fb_addViewTestsToGridPanelTest

labkey-danield committed:
Breakout SaveGridViewDialog to its own class.
Add a ManageGridViewsDialog.
Add tests for Manage Grid Views.

labkey-danield committed:
Merge branch 'develop' into fb_addViewTestsToGridPanelTest

labkey-danield committed:
Removing password for otherUser, not needed for the test.

labkey-danield committed:
Added method to remove a list of fields from the view.
Accounting for different text for grid header buttons on an app grid.
Added a test to validate default view is shown in parent samples tab.
Removed some unused imports.

labkey-chrisj reviewed (approved):
Thank you for taking on this work!

cnathe commented on src/org/labkey/test/tests/component/GridPanelViewTest.java:
```diff
@@ -0,0 +1,1389 @@
+    // Tests that need to be written:
+    // Validate "Save As..." from the grid save button.
+    // Validate views that are locked, or in some other way, cannot be updates in the manage views dialog.
+    // ???
```
do you have this list somewhere to fill out this ???

labkey-danield replied:
I've added a story to Samurai Puffin back log to track this (Grid Views -Selenium Test Part 2).
It includes these test and a couple of others.

cnathe commented on src/org/labkey/test/tests/component/GridPanelViewTest.java:
```diff
@@ -0,0 +1,1389 @@
+     *         <li>Modify the default view for a sample type in LabKey Server.</li>
+     *         <li>Save the default view, but do not make it public (default for all).</li>
+     *         <li>Validate in the App grid that 'My Default' is shown under the Views menu.</li>
+     *         <li>Validate that other users se no change to the default view.</li>
```
```suggestion
     *         <li>Validate that other users see no change to the default view.</li>
```

cnathe reviewed (commented):
The test cases here in GridPanelViewTest look great. I like that these are done via the core-components page. We talked about getting some set of these tests moved to the BVT level in some way. This would be a good one to include in that set.

labkey-tchad commented on src/org/labkey/test/components/ui/grids/CustomizeGridDialog.java:
```diff
@@ -0,0 +1,541 @@
+                {
+                    WebElement spinner = BootstrapLocators.loadingSpinner.findWhenNeeded(this);
+                    return !spinner.isDisplayed() &&
+                            elementCache().contentPanelLocator.findElements(this).size() == 2;
```
This `findWhenNeeded` -> `isDisplayed` pattern is pretty common so I recently added an `isDisplayed` method to `Locator`.
```suggestion
                    return !BootstrapLocators.loadingSpinner.isDisplayed(this) &&
                            elementCache().contentPanelLocator.findElements(this).size() == 2;
```

labkey-danield replied:
Updated.

labkey-tchad commented on src/org/labkey/test/components/ui/grids/CustomizeGridDialog.java:
```diff
@@ -0,0 +1,541 @@
+     * Check or uncheck the 'Show all system and user-defined fields' checkbox checked?
+     *
+     * @param checked Set to true to check the box, false to uncheck it.
+     * @return True if checked false otherwise.
```
Copy/paste errors
```suggestion
     * Check or uncheck the 'Show all system and user-defined fields' checkbox.
     *
     * @param checked Set to true to check the box, false to uncheck it.
     * @return This dialog
```

labkey-danield replied:
Updated.

labkey-tchad commented on src/org/labkey/test/components/ui/grids/CustomizeGridDialog.java:
```diff
@@ -0,0 +1,541 @@
+    public List<String> getAvailableFields()
+    {
+        List<WebElement> listItemElements = elementCache().getListItemNameElements(elementCache().availableFieldsPanel);
+        return listItemElements.stream().map(WebElement::getText).collect(Collectors.toList());
```
IntelliJ always suggest simplifying usages of `Collectors.toList`
```suggestion
        return listItemElements.stream().map(WebElement::getText).toList();
```

labkey-danield replied:
Yes, but that makes it an immutable list. I would stumble over that in tests that would like to add/remove elements from the list to use for comparisons.

labkey-tchad commented on src/org/labkey/test/components/ui/grids/CustomizeGridDialog.java:
```diff
@@ -0,0 +1,541 @@
+     * Expand a field or a hierarchy of fields. If a single field is passed in only it will be expanded. If multiple values
+     * are passed in it is assumed to be a path and all fields will be expanded to the last field.
+     *
+     * @param fields The list of fields to expand.
+     * @return This dialog.
+     */
+    public CustomizeGridDialog expandAvailableFields(String... fields)
+    {
+        StringBuilder fieldKey = new StringBuilder();
+
+        Iterator<String> iterator = Arrays.stream(fields).iterator();
+
+        while(iterator.hasNext())
+        {
+            fieldKey.append(iterator.next().replace(" ", ""));
+
+            expandOrCollapseByFieldKey(fieldKey.toString(), true);
+
+            // If this isn't the last item in the collection don't add a "/" to the expected data-fieldkey value.
+            if(iterator.hasNext())
+            {
+                fieldKey.append("/");
+            }
+
+        }
+
+        return this;
+    }
```
This iteration is duplicated in several methods (e.g. `addAvailableFieldToGrid` and `isAvailableFieldAddedToGrid`).
Consider refactoring this so that it can be shared. Also, not sure why it needs to be public.
```suggestion
    /**
     * Expand a field or a hierarchy of fields. If a single field is passed in only it will be expanded. If multiple values
     * are passed in it is assumed to be a path and all fields will be expanded to the last field.
     *
     * @param fields The list of fields to expand.
     * @return Row element the final specified field
     */
    public WebElement expandAvailableFields(String... fields)
    {
        if (fields.length == 0)
        {
            throw new IllegalArgumentException("Specify a field or fields to expand");
        }

        StringBuilder fieldKey = new StringBuilder();

        Iterator<String> iterator = Arrays.stream(fields).iterator();

        while(iterator.hasNext())
        {
            fieldKey.append(iterator.next().replace(" ", ""));

            expandOrCollapseByFieldKey(fieldKey.toString(), true);

            // If this isn't the last item in the collection don't add a "/" to the expected data-fieldkey value.
            if(iterator.hasNext())
            {
                fieldKey.append("/");
            }

        }

        return elementCache().getListItemElementByFieldKey(fieldKey.toString());
    }
```

labkey-danield replied:
Changed `isAvailableFieldAdded` and `addAvailableField` to call `expandAvailableFields`.
Changed `expandAvailableFields` to private. Initially was public to help some test automation that wasn't needed.

labkey-tchad commented on src/org/labkey/test/components/ui/grids/CustomizeGridDialog.java:
```diff
@@ -0,0 +1,541 @@
+
+        WebElement expandIcon = Locator.tagWithClass("div", "field-expand-icon")
+                .withChild(Locator.tagWithClass("i", iconClass))
+                .findElement(listItem);
+
+        String errorMessage;
+
+        if(expand)
+            errorMessage = String.format("There is no expand icon for field with data-fieldkey attribute '%s' in the 'Available Fields' panel.", fieldKey);
+        else
+            errorMessage = String.format("There is no collapse icon for field with data-fieldkey attribute '%s' in the 'Available Fields' panel.", fieldKey);
+
+        Assert.assertTrue(errorMessage, listItem.isDisplayed());
+
+        expandIcon.click();
+
```
This won't work correctly for fields that are already in the desired state. It should no-op for fields that are already in the expected state.
```suggestion
        if (isFieldKeyExpanded(listItem) != expand)
        {
            WebElement expandIcon = Locator.tagWithClass("div", "field-expand-icon").findElement(listItem);
            expandIcon.click();
            
            WebDriverWrapper.waitFor(() -> isFieldKeyExpanded(listItem) == expand,
                    String.format("Field, %s, did not %s.", fieldKey, expand ? "expand" : "collapse"), 5_000);
        }
```

labkey-danield replied:
I've added a NoOp check at the start that will return if the given row is already in the desired state.

labkey-tchad commented on src/org/labkey/test/components/ui/grids/CustomizeGridDialog.java:
```diff
@@ -0,0 +1,541 @@
+     * Private helper to see if the row with the given data-fieldkey value has a checkmark or not.
+     *
+     * @param fieldKey The data-fieldkey value of the row/field to check.
+     * @return True if row is expanded.
+     */
+    private boolean isFieldKeyExpanded(String fieldKey)
+    {
+        WebElement listItem = elementCache().getListItemElementByFieldKey(fieldKey);
+
+        // As long as there is no plus/expand icon then this field is expanded.
+        return   !Locator.tagWithClass("div", "field-expand-icon")
+                .withChild(Locator.tagWithClass("i", "fa-plus-square"))
+                .findWhenNeeded(listItem).isDisplayed();
+    }
```
This should probably take a `WebElement` instead of the fieldKey. Most usages of it have already found the list item before calling this. Also, fail if the field isn't expandable.
```suggestion
    /**
     * Private helper to see if the given row is expanded or not. Will throw if the row is not expandable.
     *
     * @param listItem The available field list item to inspect.
     * @return True if row is expanded.
     */
    private boolean isFieldKeyExpanded(WebElement listItem)
    {
        WebElement expandIcon = Locator.byClass("div", "field-expand-icon").childTag("i").findElement(listItem);
        return expandIcon.getAttribute("class").contains("fa-minus-square");
    }
```

labkey-danield replied:
Changed to take a WebElement. Using a WebElement made the NoOp check in expandOrCollapseByFieldKey easier.

labkey-tchad commented on src/org/labkey/test/components/ui/grids/CustomizeGridDialog.java:
```diff
@@ -0,0 +1,541 @@
+
+        // Select the current text, type in the new value then change focus (tab) to commit the change.
+        Actions replaceCurrentText = new Actions(getDriver());
+        replaceCurrentText.sendKeys(Keys.END)
+                .keyDown(Keys.SHIFT)
+                .sendKeys(Keys.HOME)
```
Just an FYI. `END` and `HOME` can behave differently on different OSs. The default behavior on MacOS is to jump to the start/end of the whole document; Linux and Windows stay on the same line (I think IntelliJ does something to mimic this behavior).
Not a problem here, since it isn't a multi-line input.

Also, there is an existing helper for this sort of scenario, `getWrapper().actionClear` will clear an input without losing focus. Note that you still can't use `setFormElement`. (I have been tiptoeing around this oddness in my `ReactInput` refactor)

labkey-danield replied:
Changed to call  `getWrapper().actionClear(elementCache().columnLabelEdit);` before setting the text using selenium Actions.

labkey-tchad commented on src/org/labkey/test/components/ui/grids/CustomizeGridDialog.java:
```diff
@@ -0,0 +1,541 @@
+     */
+    public boolean isUndoEditsEnabled()
+    {
+        // It looks like .isEnabled is not accurate if disabled-action-text class attribute is used to disable a tag.
```
We should start pushing back on this.
HTML has a `disabled` attribute that Selenium recognizes; not sure why folks don't like using it.

labkey-danield replied:
It would be nice to have some more consistency. Maybe there was some particular reason this wasn't done.

labkey-tchad commented on src/org/labkey/test/components/ui/grids/CustomizeGridDialog.java:
```diff
@@ -0,0 +1,541 @@
+/**
+ * Wraps EntityFieldFilterModal.tsx in UI components.
+ */
+public class CustomizeGridDialog extends ModalDialog
```
Wrong product component (Copy/paste error I presume). Class name should match also.
```suggestion
/**
 * Wraps CustomizeGridViewModal.tsx in UI components.
 */
public class CustomizeGridViewDialog extends ModalDialog
```

labkey-tchad commented on src/org/labkey/test/components/ui/grids/ManageGridViewsDialog.java:
```diff
@@ -0,0 +1,174 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Dialog used to manage saved views.
+ */
+public class ManageGridViewsDialog extends ModalDialog
```
The name should match the product component.
```suggestion
/**
 * Dialog used to manage saved views. Wraps ManageViewsModal.tsx in UI components.
 */
public class ManageViewsDialog extends ModalDialog
```

labkey-danield replied:
Updated.

labkey-tchad commented on src/org/labkey/test/components/ui/grids/ManageGridViewsDialog.java:
```diff
@@ -0,0 +1,174 @@
+ */
+public class ManageGridViewsDialog extends ModalDialog
+{
+    QueryGrid grid;
```
Final?
```suggestion
    private final QueryGrid grid;
```

labkey-danield replied:
Removed this variable it is not needed/used.

labkey-tchad commented on src/org/labkey/test/components/ui/grids/ManageGridViewsDialog.java:
```diff
@@ -0,0 +1,174 @@
+    {
+        elementCache().deleteNoButton.click();
+        return this;
+    }
```
These method names are confusing. What about confirm and cancel
```suggestion
    public ManageGridViewsDialog confirmDelete()
    {
        elementCache().deleteYesButton.click();
        return this;
    }

    /**
     * Click 'No' button to not delete a button.
     *
     * @return This dialog.
     */
    public ManageGridViewsDialog cancelDelete()
    {
        elementCache().deleteNoButton.click();
        return this;
    }
```

labkey-danield replied:
Changed.

labkey-tchad commented on src/org/labkey/test/components/ui/grids/QueryGrid.java:
```diff
@@ -299,13 +355,181 @@ public QueryGrid clearAllSelections()
+        }
+        else
+        {
+            return null;
```
Is there a way to wait for the save to complete? Maybe wait for the save button to disappear?
```suggestion
            getWrapper().shortWait().until(ExpectedConditions.stalenessOf(saveButton));
            return null;
```

labkey-danield replied:
Changed.

labkey-tchad commented on src/org/labkey/test/components/ui/grids/QueryGrid.java:
```diff
@@ -299,13 +355,181 @@ public QueryGrid clearAllSelections()
+     */
+    public boolean isSaveButtonVisible()
+    {
+        return Locator.buttonContainingText("Save").findWhenNeeded(elementCache().panelHeader()).isDisplayed();
```
`Locator.isDisplayed`
```suggestion
        return Locator.buttonContainingText("Save").isDisplayed(elementCache().panelHeader());
```

labkey-danield replied:
Changed.

labkey-tchad commented on src/org/labkey/test/components/ui/grids/SaveGridViewDialog.java:
```diff
@@ -0,0 +1,120 @@
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+
+/**
+ * Dialog used to save a grid view in an app.
+ */
+public class SaveGridViewDialog extends ModalDialog
```
Reference product components.
```suggestion
/**
 * Dialog used to save a grid view in an app. Wraps SaveViewModal.tsx in UI Components.
 */
public class SaveViewDialog extends ModalDialog
```

labkey-danield replied:
Changed.

labkey-tchad commented on src/org/labkey/test/tests/component/GridPanelBaseTest.java:
```diff
@@ -0,0 +1,151 @@
+        _containerHelper.createProject(getProjectName(), null);
+
+        // Add the 'Sample Types' web part. It is easier when debugging etc...
+        PortalHelper portalHelper = new PortalHelper(this);
+        portalHelper.enterAdminMode();
+        portalHelper.addWebPart("Sample Types");
+        portalHelper.exitAdminMode();
```
`addWebPart` will enter and exit admin mode if needed.
```suggestion
        new PortalHelper(this).addWebPart("Sample Types");
```

labkey-danield replied:
Changed.

labkey-tchad commented on src/org/labkey/test/tests/component/GridPanelBaseTest.java:
```diff
@@ -0,0 +1,151 @@
+    }
+
+    /**
+     * Helper to remove the 'Flag' column from the default view. It just gets in the way for some tests, and is easier
+     * to remove it.
```
How does the `Flag` column interfere?

labkey-danield replied:
It is not shown in the Apps.

labkey-tchad commented on src/org/labkey/test/tests/component/GridPanelBaseTest.java:
```diff
@@ -0,0 +1,151 @@
+        cv.removeColumn(REMOVED_FLAG_COLUMN);
+
+        // This will save the changes to the default view for everyone. The view menu option will remain as "Default"
+        cv.saveCustomView("", true);
```
Could probably safely make `CustomizeView.saveDefaultView` default to `shared=true`, or add a new `saveDefaultView` override that takes the shared flag.

labkey-danield replied:
Added and override `CustomizeView.saveDefaultView(boolean)`.

labkey-tchad commented on src/org/labkey/test/tests/component/GridPanelBaseTest.java:
```diff
@@ -0,0 +1,151 @@
+        QueryGrid grid = CoreComponentsTestPage.beginAt(this, getProjectName())
+                .getGridPanel(TEST_SCHEMA, sampleType);
+
+        grid.clearAllSelections();
+
+        if(clearFilters)
+        {
+            grid.clearFilters();
+        }
+
+        grid.clearSearch();
```
Consider adding a `resetView` method to `QueryGrid` that does all of this. This method has no way to deal with a combination of custom filters and view filters.
```suggestion
        grid.resetView();
```
In `QueryGrid`:
```
    public void resetView()
    {
        clearAllSelections();

        if(isUndoButtonVisible())
        {
            clickUndoButton();
        }

       clearSearch();
    }
```

labkey-danield replied:
I see your point, but can you think of a better method name? Now that views are available in this grid, and you can edit an existing view, a method named `resetView` could be misinterpreted.
Maybe something more verbose like `resetToDefaultAndClearSelections()` or maybe just `resetToDefaultState()`?

labkey-tchad replied:
Oh, certainly. I wasn't attached to the name; just thought it would be useful functionality since we frequently want to just reset a grid.

labkey-danield replied:
I've `added QueryGrid.resetToDefaultState(boolean revertDefaultView)`.
It will do the best it can to get the grid back to its original state. It will apply the default view. If the parameter is set to true it will revert the default view before applying it, otherwise it will apply the default view with any modifications it may have.

labkey-tchad commented on src/org/labkey/test/tests/component/GridPanelViewTest.java:
```diff
@@ -0,0 +1,1389 @@
+
+        checker().screenShotIfNewError(String.format("%s_Views_Menu_Column_Error", screenShotPrefix));
+
+        return checker().errorsSinceMark() == 0;
```
`screenshotIfNewError` resets the error mark. This will always return `true`.
I think it would make sense for `screenshotIfNewError` to return `true` if it took a screenshot.

labkey-danield replied:
I've changed the DeferredErrorCollector.screenShotIfNewError to return a boolean. True if it took a screenshot, false otherwise.

labkey-tchad commented on src/org/labkey/test/tests/component/GridPanelViewTest.java:
```diff
@@ -0,0 +1,1389 @@
+
+        checker().screenShotIfNewError(String.format("%s_Grid_Header_Error", screenShotPrefix));
+
+        return checker().errorsSinceMark() == 0;
```
Always `true`

labkey-danield replied:
Same comment as above.

labkey-danield committed:
Add code to click the tab for the sample grid on the Storage View page.
Added gridview test (remove columns in default view) for LKSM.
Moved some shared test code to static method in SMTestUtils.
Some code clean up and class renaming.

labkey-danield committed:
Merge branch 'develop' into fb_addViewTestsToGridPanelTest

labkey-danield committed:
Renamed method for clarity.
Added methods to the ManageGridViewsDialog that are only available in apps.
Added test for filtering the default view.
Consolidated some shared test code.
Added test for Sample Finder to the Remove Column Default View tests.

labkey-danield committed:
Changed locators in ManageGridViewsDialog.
Moved shared test code to base class.
Added test that validate app specific features of the Manage Views Dialog.

labkey-danield committed:
Merge branch 'develop' into fb_addViewTestsToGridPanelTest

labkey-danield committed:
Merge branch 'develop' into fb_addViewTestsToGridPanelTest

labkey-danield committed:
Changed most of the grid view tests in BiologicsComponentTest to use the in app grid view functionality.
Added some methods to CustomizeGridDialog, ManageGridViewsDialog and QueryGrid to that do multiple procedural steps.
Updated some method names to be more like customize view test helper in LKS.

labkey-danield committed:
Merge branch 'develop' into fb_addViewTestsToGridPanelTest

labkey-danield committed:
Updated CrossFolderRegistryTest to use grid view component in the app.
Updated saveView method in grid to have an overload that will allow setting the makeAvailable option.

labkey-danield committed:
Some more grid view updates in LKB.
Added overload to save view as default.

labkey-danield committed:
Merge branch 'develop' into fb_addViewTestsToGridPanelTest

labkey-danield committed:
Merge branch 'develop' into fb_addViewTestsToGridPanelTest

labkey-danield committed:
Add test for cross folder views.

labkey-danield committed:
Merge branch 'develop' into fb_addViewTestsToGridPanelTest

labkey-danield committed:
Update src/org/labkey/test/tests/component/GridPanelViewTest.java

Co-authored-by: Cory Nathe <cnathe@labkey.com>

labkey-danield committed:
Update src/org/labkey/test/components/ui/grids/CustomizeGridDialog.java

Co-authored-by: Trey Chadick <tchad@labkey.com>

labkey-danield committed:
Update src/org/labkey/test/components/ui/grids/CustomizeGridDialog.java

Co-authored-by: Trey Chadick <tchad@labkey.com>

labkey-danield committed:
Implementing code review feedback.

labkey-danield committed:
Merge remote-tracking branch 'origin/fb_addViewTestsToGridPanelTest' into fb_addViewTestsToGridPanelTest

# Conflicts:
#	src/org/labkey/test/components/ui/grids/CustomizeGridDialog.java

labkey-danield committed:
Merge branch 'develop' into fb_addViewTestsToGridPanelTest

labkey-danield committed:
Update src/org/labkey/test/components/ui/grids/CustomizeGridDialog.java

Co-authored-by: Trey Chadick <tchad@labkey.com>

labkey-danield committed:
Implementing code review feedback.
Changed DeferredErrorCollector.screenShotIfNewError to return true if a screenshot was taken.
Added CustomizeView.saveDefaultView(bbolean) method to give the option of making the default view shared.

labkey-danield committed:
Reversing return value for screenShotIfNewError.

labkey-tchad commented on src/org/labkey/test/util/DeferredErrorCollector.java:
```diff
@@ -156,17 +156,21 @@ public int errorsSinceMark()
      * @see #takeScreenShot(String)
      */
-    public void screenShotIfNewError(@NotNull String screenshotName)
+    public boolean screenShotIfNewError(@NotNull String screenshotName)
     {
+        boolean noError = true;
         if (errorsSinceMark() > 0)
         {
             final String s = artifactCollector.dumpPageSnapshot(screenshotName);
             allErrors.get(allErrors.size() - 1).setScreenshotName(s);
             artifactCollector.reportTestMetadata(s);
+            noError = false;
         }
         setErrorMark();
+        return noError;
```
The `setErrorMark` call is a no-op if there aren't any errors so the variable isn't really necessary here (not sure why I put the call outside the if).
I'd also flip the return value; if I saw `if (checker().screenShotIfNewError())`, I would expect it to return `true` if it took a screenshot.
```suggestion
    public boolean screenShotIfNewError(@NotNull String screenshotName)
    {
        if (errorsSinceMark() > 0)
        {
            final String s = artifactCollector.dumpPageSnapshot(screenshotName);
            allErrors.get(allErrors.size() - 1).setScreenshotName(s);
            artifactCollector.reportTestMetadata(s);
            setErrorMark();
            return true;
        }
        return false;
```

labkey-danield replied:
Changed to return true if a screenshot is taken. Updated comment as well.

labkey-tchad reviewed (approved):
Just a couple of small changes

labkey-danield committed:
Merge branch 'develop' into fb_addViewTestsToGridPanelTest

labkey-danield committed:
Have screenShotIfNewError return true if it took a screenshot.
Added QueryGrid.resetToDefaultState to reset a grid to a 'default' state.
